L1Trigger/TrackerTFP/src/HoughTransform.cc

```diff
@@ -0,0 +1,200 @@
+    auto validStub = [](int& sum, StubGP* stub){ return sum += stub ? 1 : 0; };
+    int nStubsHT(0);
+    for (const vector<deque<StubGP*>>& binInv2R : input_)
+      for (const deque<StubGP*>& sector : binInv2R)
+        nStubsHT += accumulate(sector.begin(), sector.end(), 0, validStub);
+    stubsHT_.reserve(nStubsHT);
+  }
+
+  // fill output products
+  void HoughTransform::produce(TTDTC::Streams& accepted, TTDTC::Streams& lost) {
+    for (int binInv2R = 0; binInv2R < dataFormats_->numChannel(Process::ht); binInv2R++) {
+      const int inv2R = inv2R_.toSigned(binInv2R);
+      deque<StubHT*> acceptedAll;
+      deque<StubHT*> lostAll;
+      for (deque<StubGP*>& inputSector : input_[binInv2R]) {
+        const int size = inputSector.size();
+        vector<StubHT*> acceptedSector;
+        vector<StubHT*> lostSector;
+        acceptedSector.reserve(size);
+        lostSector.reserve(size);
+        // associate stubs with inv2R and phiT bins
+        fillIn(inputSector, acceptedSector, lostSector, inv2R);
+        // Process::ht collects all stubs before readout starts -> remove all gaps
+        acceptedSector.erase(remove(acceptedSector.begin(), acceptedSector.end(), nullptr), acceptedSector.end());
+        // identify tracks
+        readOut(acceptedSector, lostSector, acceptedAll, lostAll);
+      }
+      // truncate accepted stream
+      const auto limit = enableTruncation_ ? next(acceptedAll.begin(), min(setup_->numFrames(), (int)acceptedAll.size())) : acceptedAll.end();
+      copy_if(limit, acceptedAll.end(), back_inserter(lostAll), [](StubHT* stub){ return stub; });
+      acceptedAll.erase(limit, acceptedAll.end());
+      // store found tracks
+      auto put = [](const deque<StubHT*>& stubs, TTDTC::Stream& stream){
+        stream.reserve(stubs.size());
+        for (StubHT* stub : stubs)
+          stream.emplace_back(stub ? stub->frame() : TTDTC::Frame());
+      };
+      const int offset = region_ * dataFormats_->numChannel(Process::ht);
+      put(acceptedAll, accepted[offset + binInv2R]);
+      // store lost tracks
+      put(lostAll, lost[offset + binInv2R]);
+    }
+  }
+
+  // associate stubs with inv2R and phiT bins
+  void HoughTransform::fillIn(deque<StubGP*>& inputSector, vector<StubHT*>& acceptedSector, vector<StubHT*>& lostSector, int inv2R) {
+    // fifo, used to store stubs which belongs to a second possible track
+    deque<StubHT*> stack;
+    // clock accurate firmware emulation, each while trip describes one clock tick, one stub in and one stub out per tick
+    while (!inputSector.empty() || !stack.empty()) {
+      StubHT* stubHT = nullptr;
+      StubGP* stubGP = pop_front(inputSector);
+      if (stubGP) {
+        const double phiT = stubGP->phi() - inv2R_.floating(inv2R) * stubGP->r();
+        const int major = phiT_.integer(phiT);
+        if (phiT_.inRange(major)) {
+          // major candidate has pt > threshold (3 GeV)
+          stubsHT_.emplace_back(*stubGP, major, inv2R);
```

Comment:
Add comment "stubHT records which HT bin this stub is added to".

Reply:
done